src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -160,35 +175,443 @@ private DeclaredTypeAssignment GetParameterType(ParameterDeclarationSyntax synta
+                        objectName.Append('\'');
+                    }
+
+                    if (prop.OptionalityMarker is not null)
+                    {
+                        objectName.Append('?');
+                    }
+
+                    objectName.Append(": ");
+                    objectName.Append(GetPropertyTypeName(prop.Value, propertyType));
+                    objectName.Append(", ");
+
+                    var propertyFlags = prop.OptionalityMarker is null ? TypePropertyFlags.Required : TypePropertyFlags.None;
+                    properties.Add(new(propertyName, propertyType, propertyFlags, TryGetTypePropertyDescription(prop)));
+                } else
+                {
+                    diagnostics.Add(DiagnosticBuilder.ForPosition(prop.Key).NonConstantTypeProperty());
+                    // since we're not attaching this property to the object due to the non-constant key, forward any property errors to the object type
+                    if (propertyType is ErrorType error)
+                    {
+                        diagnostics.AddRange(error.GetDiagnostics());
+                    }
+                }
+            }
+
+            if (diagnostics.Any())
+            {
+                // foward any diagnostics gathered from parsing properties to the return type. normally, these diagnostics would be gathered by the SemanticDiagnosticVisitor (which would visit the properties of an ObjectType looking for errors).
+                // Errors hidden behind DeferredTypeReferences will unfortunately be dropped, as we can't call their type function without risking an infinite loop (in the case that a recursive object type has errors)
+                return ErrorType.Create(diagnostics.Concat(properties.Select(p => p.TypeReference).OfType<TypeSymbol>().SelectMany(e => e.GetDiagnostics())));
+            }
+
+            objectName.Remove(objectName.Length - 2, 1);
+            objectName.Append('}');
+
+            var parentSyntax = binder.GetParent(syntax);
+
+            var typeFlags = UnwrapUntilDecorable(syntax, HasSecureDecorator, TypeSymbolValidationFlags.IsSecure, TypeSymbolValidationFlags.Default);
+
+            (TypeSymbol? type, TypePropertyFlags flags) additionalProperties
+                = UnwrapUntilDecorable(syntax, HasSealedDecorator, (null, TypePropertyFlags.None), (LanguageConstants.Any, TypePropertyFlags.FallbackProperty));
+
+            return new ObjectType(objectName.ToString(), typeFlags, properties, additionalProperties.type, additionalProperties.flags);
+        }
+
+        private string GetPropertyTypeName(SyntaxBase typeSyntax, ITypeReference propertyType)
+        {
+            if (typeSyntax is TypeAccessSyntax typeAccess)
+            {
+                return typeAccess.Name.IdentifierName;
+            }
+
+            return propertyType switch
+            {
+                DeferredTypeReference deferred => "<deferred>",
```

Comment:
Is this to avoid recursion? Could you add a comment to explain?

If so, won't the recursive case always have a named type to refer to?